components/event-sources/adapter/http/adapter.go

```diff
@@ -0,0 +1,196 @@
+
+	// note about graceful shutdown:
+	// TLDR; StartReceiver unblocks as soon as a stop signal is received
+	// `StartReceiver` waits internally until `ctx.Done()` does not block anymore
+	// the context `h.adapterContext` returns a channel (when calling `ctx.Done()`)
+	// which is closed as soon as a stop signal is received, see https://github.com/knative/pkg/blob/master/signals/signal.go#L37
+	if err := c.StartReceiver(h.adapterContext, h.serveHTTP); err != nil {
+		return fmt.Errorf("failed to start receiver: %v", err)
+	}
+	logger.Info("adapter stopped")
+	return nil
+}
+
+type readinessMiddleware struct {
+	handler http.Handler
+}
+
+func WithReadinessMiddleware(next http.Handler) http.Handler {
+	return readinessMiddleware{handler: next}
+}
+
+func (r readinessMiddleware) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if req.URL.Path == readinessReadiness {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+	r.handler.ServeHTTP(w, req)
+}
+
+// serveHTTP handles incoming events
+// enriches them with the source
+// and sends them to the sink
+// NOTE: EventResponse reason is empty since it is ignored: https://github.com/cloudevents/sdk-go/blob/master/pkg/cloudevents/transport/http/transport.go#L496
+// NOTE: instead the error is used as error message for request response
+func (h *httpAdapter) serveHTTP(ctx context.Context, event cloudevents.Event, resp *cloudevents.EventResponse) error {
+	logger := h.logger
+	logger.Debug("got event", zap.Any("event_context", event.Context))
+
+	if !isSupportedCloudEvent(event) {
+		resp.Error(http.StatusBadRequest, "")
+		return errors.New(ErrorResponseCEVersionUnsupported)
+	}
+
+	// validate event conforms to cloudevents specification
+	if err := event.Validate(); err != nil {
+		resp.Error(http.StatusBadRequest, "")
+		return err
+	}
+
+	// enrich the event with the application source
+	// the application source is injected into this adapter from the http source controller
+	event.SetSource(h.accessor.GetSource())
+
+	reportArgs := &source.ReportArgs{
+		Namespace:     h.accessor.GetNamespace(),
+		EventSource:   event.Source(),
+		EventType:     event.Type(),
+		Name:          "http_adapter",
+		ResourceGroup: resourceGroup,
+	}
+
+	logger.Debug("sending event", zap.Any("sink", h.accessor.GetSinkURI()))
+	rctx, revt, err := h.ceClient.Send(ctx, event)
+	if err != nil {
+		h.logger.Error("failed to send cloudevent to sink", zap.Error(err), zap.Any("sink", h.accessor.GetSinkURI()))
+		resp.Error(http.StatusBadGateway, "")
+		// do not show this error to user, might contain sensitive information
+		return nil
+	}
+
+	rtctx := cloudevents.HTTPTransportContextFrom(rctx)
+	if rtctx.StatusCode == 0 {
+		resp.RespondWith(http.StatusInternalServerError, revt)
+		return nil
+	}
+
+	// report a sent event
+	if err := h.statsReporter.ReportEventCount(reportArgs, rtctx.StatusCode); err != nil {
+		h.logger.Warn("cannot report event count", zap.Error(err))
+	}
+
+	if rtctx.StatusCode/100 == 2 {
```

Comment:
wat

Reply:
Ideally there would be a helper in `http`. `http.is2XX(statusCode)` which would improve readability. Let me check if there is something like this.

Reply:
Did not find a nice helper. Refactored it to a function to improve readability.